src/server/frontend_wayland/idle_inhibit_v1.cpp

```diff
@@ -0,0 +1,173 @@
+/*
+ * Copyright © 2022 Canonical Ltd.
+ *
+ * This program is free software: you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 3,
+ * as published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ * Authored by: Grayson Guarino <grayson.guarino@canonical.com>
+ */
+
+#include "idle_inhibit_v1.h"
+
+#include "mir/log.h"
+#include "mir/executor.h"
+#include "mir/scene/idle_hub.h"
+
+#include <boost/throw_exception.hpp>
+#include <memory>
+
+namespace mf = mir::frontend;
+namespace mw = mir::wayland;
+namespace ms = mir::scene;
+
+namespace mir
+{
+namespace frontend
+{
+struct IdleInhibitV1Ctx
+{
+    std::shared_ptr<Executor> const wayland_executor;
+    std::shared_ptr<scene::IdleHub> const idle_hub;
+};
+
+class IdleInhibitManagerV1Global
+    : public wayland::IdleInhibitManagerV1::Global
+{
+public:
+    IdleInhibitManagerV1Global(wl_display* display, std::shared_ptr<IdleInhibitV1Ctx> const& ctx);
+
+private:
+    void bind(wl_resource* new_resource) override;
+
+    std::shared_ptr<IdleInhibitV1Ctx> const ctx;
+};
+
+class IdleInhibitManagerV1
+    : public wayland::IdleInhibitManagerV1
+{
+public:
+    IdleInhibitManagerV1(wl_resource* resource, std::shared_ptr<IdleInhibitV1Ctx> const& ctx);
+
+private:
+    void create_inhibitor(struct wl_resource* id, struct wl_resource* surface) override;
+
+    std::shared_ptr<IdleInhibitV1Ctx> const ctx;
+};
+
+class IdleInhibitorV1
+    : public wayland::IdleInhibitorV1
+{
+public:
+    IdleInhibitorV1(wl_resource* resource, std::shared_ptr<IdleInhibitV1Ctx> const& ctx);
+    ~IdleInhibitorV1();
+
+private:
+    // Not sure why, but removing this unused StateObserver stops the functionality of the IdleInhibit.
+    // It seems unlikely that it will ever actually be used.
```

Comment:
Not sure why, but this comment suggest that this class isn't used when it clearly is.

Reply:
I never need to use the `StateObserver` for anything here, but when it is removed, Idle Inhibit stops functioning. Is there some behavior that `StateObserver` is doing that I am unaware of?

Reply:
```C++
          state_observer{std::make_shared<StateObserver>(this)}
{
//...
    ctx->idle_hub->register_interest(state_observer, time::Duration(0));
```